src/main/java/com/okta/tools/saml/OktaSaml.java

```diff
@@ -50,22 +50,46 @@ public String getSamlResponse() throws IOException {
 
     private String getSamlResponseForAws(String oktaSessionToken) throws IOException {
         Document document = launchOktaAwsAppWithSessionToken(environment.oktaAwsAppUrl, oktaSessionToken);
-        Elements samlResponseInputElement = document.select("form input[name=SAMLResponse]");
-        if (samlResponseInputElement.isEmpty()) {
-            throw new RuntimeException("You do not have access to AWS through Okta. \nPlease contact your administrator.");
-        }
-        return samlResponseInputElement.attr("value");
+        return getSamlResponseForAwsFromDocument(document);
     }
 
     private String getSamlResponseForAwsRefresh() throws IOException {
         Document document = launchOktaAwsApp(environment.oktaAwsAppUrl);
+        return getSamlResponseForAwsFromDocument(document);
+    }
+
+    private String getSamlResponseForAwsFromDocument(Document document) {
         Elements samlResponseInputElement = document.select("form input[name=SAMLResponse]");
         if (samlResponseInputElement.isEmpty()) {
-            throw new RuntimeException("You do not have access to AWS through Okta. \nPlease contact your administrator.");
+            if (isPasswordAuthenticationChallenge(document)) {
+                throw new IllegalStateException("Unsupported App sign on rule: 'Prompt for re-authentication'. \nPlease contact your administrator.");
+            } else if (isPromptForFactorChallenge(document)) {
+                throw new IllegalStateException("Unsupported App sign on rule: 'Prompt for factor'. \nPlease contact your administrator.");
+            } else {
+                Elements errorContent = document.getElementsByClass("error-content");
+                Elements errorHeadline = errorContent.select("h1");
+                if (errorHeadline.hasText()) {
+                    throw new RuntimeException(errorHeadline.text());
+                } else {
+                    throw new RuntimeException("You do not have access to AWS through Okta. \nPlease contact your administrator.");
+                }
+            }
         }
         return samlResponseInputElement.attr("value");
     }
 
+    // Heuristic based on undocumented behavior observed experimentally
+    // This condition may be missed Okta significantly changes the app-level re-auth page
+    private boolean isPasswordAuthenticationChallenge(Document document) {
+        return document.getElementById("password-verification-challenge") != null;
+    }
```

Comment:
@randomsamples hopefully the comments on these methods make up for the lack of documentation. The risk is mostly that the tool will incorrectly say that the user doesn't have app access if Okta changes these pages (the current behavior prior to this change). Is that an acceptable compromise?